File: mParticle-Apple-SDK/Utils/MPIHasher.m
#import "MPIHasher.h"

@implementation MPIHasher

+ (uint64_t)hashFNV1a:(NSData *)data {
    // FNV-1a hashing
    uint64_t rampHash = 0xcbf29ce484222325;
    const char *bytes = (const char *)data.bytes;
    NSUInteger length = data.length;
    
    for (int i = 0; i < length; i++) {
        rampHash = (rampHash ^ bytes[i]) * 0x100000001B3;
    }
    return rampHash;
}

+ (NSString *)hashString:(NSString *)stringToHash {
    if (stringToHash.length == 0) {
        return @"";
    }
    
    NSString *lowercaseStringToHash = stringToHash.lowercaseString;
    NSData *dataToHash = [lowercaseStringToHash dataUsingEncoding:NSUTF8StringEncoding];
    const char *bytes = (const char *)dataToHash.bytes;
    NSUInteger length = dataToHash.length;
    
    int32_t hash = 0;
    for (int i = 0; i < length; i++) {
        uint8_t character = bytes[i];
        hash = ((hash << 5) - hash) + character;
    }
    
    return [NSString stringWithFormat:@"%d", hash];
}

+ (NSString *)hashStringUTF16:(NSString *)stringToHash {
    NSData *data = [stringToHash dataUsingEncoding:NSUTF16LittleEndianStringEncoding];
    int64_t hash = [self hashFNV1a:data];
    NSString *result = @(hash).stringValue;
    return result;
}

+ (NSString *)hashEventType:(MPEventType)eventType {
    return [self hashString:[@(eventType) stringValue]];
}

+ (MPEventType)eventTypeForHash:(NSString *)hashString {
    for (int i = 1; i <= MPEventTypeImpression; ++i) {
        if ([hashString isEqualToString: [MPIHasher hashEventType: (MPEventType)i]]) {
            return (MPEventType)i;
        }
    }
    
    return MPEventTypeOther;
}

+ (NSString *)hashEventType:(MPEventType)eventType eventName:(NSString *)eventName isLogScreen:(BOOL)isLogScreen {
    NSString *stringToBeHashed;
    if (isLogScreen) {
        stringToBeHashed = [NSString stringWithFormat:@"%@%@", @"0", [eventName lowercaseString]];

    } else {
        stringToBeHashed = [NSString stringWithFormat:@"%@%@", [@(eventType) stringValue], [eventName lowercaseString]];
    }
    
    return [self hashString:stringToBeHashed];
}

+ (NSString *)hashEventAttributeKey:(MPEventType)eventType eventName:(NSString *)eventName customAttributeName:(NSString *)customAttributeName isLogScreen:(BOOL)isLogScreen {
    NSString *stringToBeHashed;
    if (isLogScreen) {
        stringToBeHashed = [NSString stringWithFormat:@"%@%@%@", @"0", eventName, customAttributeName];
        
    } else {
        stringToBeHashed = [NSString stringWithFormat:@"%@%@%@", [@(eventType) stringValue], eventName, customAttributeName];
    }
    return [self hashString:stringToBeHashed];
}

+ (NSString *)hashUserAttributeKey:(NSString *)userAttributeKey {
    return [self hashString:userAttributeKey.lowercaseString];
}

+ (NSString *)hashUserAttributeValue:(NSString *)userAttributeValue {
    return[self hashString:userAttributeValue.lowercaseString];
}

    // User Identities are not actually hashed, this method is named this way to
    // be consistent with the filter class. UserIdentityType is also a number
+ (NSString *)hashUserIdentity:(MPUserIdentity)userIdentity {
    return [[NSString alloc] initWithFormat:@"%lu", (unsigned long)userIdentity];
}

+ (NSString *)hashConsentPurpose:(NSString *)regulationPrefix purpose:(NSString *)purpose {
    NSString *stringToBeHashed = [NSString stringWithFormat:@"%@%@", regulationPrefix, [purpose lowercaseString]];
    return [self hashString:stringToBeHashed];
}

+ (NSString *)hashCommerceEventAttribute:(MPEventType)commerceEventType key:(NSString *)key {
    NSString *stringToBeHashed = [NSString stringWithFormat:@"%@%@", [@(commerceEventType) stringValue], key];
    return [self hashString:stringToBeHashed];
}

+ (NSString *)hashTriggerEventName:(NSString *)eventName eventType:(NSString *)eventType {
    NSString *stringToBeHashed = [NSString stringWithFormat:@"%@%@", eventName, eventType];
    return [self hashString:stringToBeHashed];
}

@end

File: mParticle-Apple-SDK/Utils/MPIHasher.swift
//
//  MPIHasher.swift
//  mParticle-Apple-SDK
//
//  Created by Ben Baron on 10/24/23.
//

import Foundation

@objc public class MPIHasher : NSObject {

    @objc public class func hashFNV1a(_ data: Data) -> Int64 {
        var rampHash: UInt64 = 0xcbf29ce484222325
        
        for byte in data {
            rampHash = (rampHash ^ UInt64(byte)) &* 0x100000001B3
        }
        return Int64(bitPattern: rampHash)
    }

    @objc public class func hashString(_ stringToHash: String) -> String {
        if stringToHash.isEmpty {
            return "";
        }
        
        let lowercaseStringToHash = stringToHash.lowercased()
        guard let dataToHash = lowercaseStringToHash.data(using: .utf8) else {
            MPLog.warning("Hash String Failed. Could not encode string as data")
            return ""
        }
       
        var hash: Int32 = 0
        for byte in dataToHash {
            hash = ((hash << 5) &- hash) &+ Int32(byte);
        }
        
        return String(hash)
    }

    @objc public class func hashStringUTF16(_ stringToHash: String) -> String {
        guard let data = stringToHash.data(using: .utf16LittleEndian) else {
            MPLog.warning("Hash String UTF16 Failed. Could not encode string as data")
            return ""
        }
        let hash = hashFNV1a(data)
        return String(hash)
    }

    @objc public class func hashEventType(_ eventType: MPEventType) -> String {
        return hashString(String(eventType.rawValue))
    }

    @objc public class func eventType(forHash hashString: String) -> MPEventType {
        for i in 1...MPEventType.impression.rawValue {
            if let eventType = MPEventType(rawValue: i), hashString == hashEventType(eventType) {
                return eventType
            }
        }
        return .other
    }

    @objc public class func hashEventType(_ eventType: MPEventType, eventName: String, isLogScreen: Bool) -> String {
        let stringToBeHashed: String
        if isLogScreen {
            stringToBeHashed = "0\(eventName)"
        } else {
            stringToBeHashed = "\(eventType.rawValue)\(eventName)"
        }
        return hashString(stringToBeHashed)
    }

    @objc public class func hashEventAttributeKey(_ eventType: MPEventType, eventName: String, customAttributeName: String, isLogScreen: Bool) -> String {
        let stringToBeHashed: String
        if isLogScreen {
            stringToBeHashed = "0\(eventName)\(customAttributeName)"
        } else {
            stringToBeHashed = "\(eventType.rawValue)\(eventName)\(customAttributeName)"
        }
        return hashString(stringToBeHashed)
    }

    @objc public class func hashUserAttributeKey(_ userAttributeKey: String) -> String {
        return hashString(userAttributeKey)
    }

    @objc public class func hashUserAttributeValue(_ userAttributeValue: String) -> String {
        return hashString(userAttributeValue)
    }

    // User Identities are not actually hashed, this method is named this way to
    // be consistent with the filter class. UserIdentityType is also a number
    @objc public class func hashUserIdentity(_ userIdentity: MPUserIdentity) -> String {
        return String(userIdentity.rawValue)
    }

    @objc public class func hashConsentPurpose(_ regulationPrefix: String, purpose: String) -> String {
        return hashString("\(regulationPrefix)\(purpose)")
    }

    @objc public class func hashCommerceEventAttribute(_ commerceEventType: MPEventType, key: String) -> String {
        return hashString("\(commerceEventType.rawValue)\(key)")
    }

    @objc public class func hashTriggerEventName(_ eventName: String, eventType: String) -> String {
        return hashString("\(eventName)\(eventType)")
    }
}
